SalesforceFoundation/EDA
# [EDA] Prevent Deletion of Program Plan Record if There are Associated Child Records

# Critical Changes

# Changes
## Updates to Ensure Data Integrity
We've expanded the functionality that prevents the deletion of certain records when they have related child records. Now, the following record types cannot be deleted if they have the following related records.

- Program Plan records can't be deleted if they have related Program Enrollment or Plan Requirement records.

This functionality is for internal use only and is enabled in Hierarchy Custom Settings for these objects by default in new and existing orgs. If you've disabled any of the object _CannotDelete_TDTM classes for these objects in an existing org, they will remain disabled even if they are enabled in Hierarchy Custom Settings.
**Note:** We strongly recommend that EDA Admins don't manage Hierarchy Custom Settings directly and, instead, use the EDA Settings UI, which will be provided in an upcoming release.

# Issues Closed
Closes #1120 

# New Metadata

# Deleted Metadata

# Testing Notes
Scenario 1: (Before merged into Master)

1) Ensure  'Prevent Program Plan Deletion' is defaulted to TRUE on scratch org creation i.e., Set up -> Custom Settings -> Hierarchy Settings -> Prevent Program Plan Deletion -> Should've been set to TRUE
2) Create a Program Plan
3) Create Plan Requirement and/or Program Enrollment as child records to the Program Plan created in step 2
4) Delete the Program Plan created in step 2
5) You should receive the error message https://salesforce.quip.com/KmXZALIhSsgW#ZYaACAcCF8m
6) Now, uncheck the 'Prevent Program Plan Deletion' check box from step 1
7) Delete the Program Plan created in step 2
8) The Program plan should be deleted

Scenario 2: (After Merged into Master)
We're utilizing install script to set 'Prevent Program Plan' Hierarchy Settings to TRUE for both existing users and new users. It would be nice to test if that this is reflected in a Push upgrade. May be using upgraded_org flow. 

Note: The install script was updated assuming this would go into release 1.96. If it doesn't make into 1.96 the install script should be updated accordingly.
